### DISCUSSION TITLE
Unable to cURL auth.tesla.com

### DISCUSSION THREAD

**Discussion by alexmaple** [timdorr/tesla-api#330] [Q&A]
We have had one of our scripts break this afternoon and it looks like we are now unable to cURL https://auth.tesla.com.

It seems to work over the command line but not in our scripts for some reason, which doesn't make a lot of sense... the request seems to just hang and go on indefinitely.

Is anyone else experiencing this?

**Comment by briangru** [comment]
Try adding a User-agent string to your header.  That got me past most of my errors.

**Comment by alexmaple** [comment]
Just an update on this. I can get through to `https://auth.tesla.com/oauth2/v3/authorize` about once every hour or so using PHP curl, but then all other requests infinitely timeout.

It would appear the request is being blocked by their firewall due to a certain mix of curl options and maybe server configurations.

I have also tried this across multiple different servers (AWS and DO)

Will report back if I ever get to a solution.

**Reply by JulesStringer** [reply]
I wish you good luck.
For most of February I was successfully using a script derived from IMgoRT's version of the ts code at https://github.com/timdorr/tesla-api/discussions/263. At 8am (GMT) on Friday I got a status code 403 from step1, which persisted all day until I spotted that my battery's backup_reserve_percent wouldn't move off 66%. ( I have a gateway 1 so cannot adjust the reserve through the app).
On Sunday I started a rewrite in js under nodejs, with which I am more familiar. At the first attempt step 1 returned html which unfortunately I didn't capture. With many attempts that day it would generally hang (rather than return 403).
Monday morning I modified the code to capture the results of each stage, save all intermediate results and the stage reached.
I was able to save the html from step 1, and after a few changes has a return from step 2 which hit an error.
Since then it has always hung at either step1 or step2. 
I pinged auth.tesla.com and found it pointed to a subdomain of akamaiedge.net which belongs to akamai technologies whose products include a WAF.
I got nowhere this morning and decide that if I ever got authenticated again another problem might arise in a month's time so I rang Tesla support and eventually managed to get them to reset my backup reserve. 
I think the issue is that once you have been identified as a user the WAF wants to block it must spot some patterns in your request, so after a bit it gets near 100% effective at blocking your requests. So a non-programming solution is needed (get Tesla to release the API?/ change tariff / get battery with a published API Victron?)
As I am one retired user this is getting beyond me, and I have decided to give up on scheduling a powerwall to use Octopus agile tariff, so am now looking elsewhere.
